[pkg/util/log] Scrub 'pwd' and certificates when sending flare

KSerrania opened the pull request:
### What does this PR do?

`pwd` has been added to the list of password keys to scrub if encountered.
To remove private certificates / keys, multiline replacers have been introduced in the credentials scrubbing phase (since certificates are usually written on multiple lines).
Searching for certificates / keys is done by a regex which follows [RFC 7468](https://tools.ietf.org/html/rfc7468)'s ABNF as closely as possible. However, the lack of backreferences makes it impossible to check in the regex that the same label is used in the `BEGIN` and `END` parts.
Tests have been added for these two cases.

### Motivation

The `pwd` and `cert_key` fields of [this integration config file](https://github.com/DataDog/integrations-core/blob/c603898bbcb7a9d2801322b8353d729e54dcf6a1/cisco_aci/datadog_checks/cisco_aci/data/conf.yaml.example) weren't properly scrubbed when sending a flare.

### Benchmarks

Before change:

```
goos: darwin
goarch: amd64
pkg: github.com/DataDog/datadog-agent/pkg/util/log
BenchmarkLogVanilla-8                1000000          2297 ns/op        1068 B/op         15 allocs/op
BenchmarkLogVanillaLevels-8         30000000            57.5 ns/op        64 B/op          2 allocs/op
BenchmarkLogScrubbing-8               100000         13699 ns/op        6605 B/op         38 allocs/op
BenchmarkLogScrubbingLevels-8       20000000            71.2 ns/op        16 B/op          1 allocs/op
BenchmarkLogScrubbingMulti-8          100000         16601 ns/op        7564 B/op         54 allocs/op
PASS
coverage: 26.3% of statements
```

After change:
```
goos: darwin
goarch: amd64
pkg: github.com/DataDog/datadog-agent/pkg/util/log
BenchmarkLogVanilla-8                1000000          2243 ns/op        1068 B/op         15 allocs/op
BenchmarkLogVanillaLevels-8         30000000            58.1 ns/op        64 B/op          2 allocs/op
BenchmarkLogScrubbing-8               100000         14468 ns/op        6698 B/op         40 allocs/op
BenchmarkLogScrubbingLevels-8       20000000            71.8 ns/op        16 B/op          1 allocs/op
BenchmarkLogScrubbingMulti-8          100000         16225 ns/op        7660 B/op         56 allocs/op
PASS
coverage: 27.7% of statements
```

KSerrania committed:
[pkg/util/log] Scrub 'pwd' and certificates when sending flare

The 'pwd' has been added to the list of password keys to scrub.
To remove private certificates / keys, multiline replacers have
been introduced in the credentials scrubbing phase (since
certificates are usually written on multiple lines).
Tests have been added for these two cases.

olivielpeau commented:
haven't reviewed the changes by themselves, a few notes/pointers though :) :

* `CredentialsCleanerBytes` is used by the logger on every log message that's logged, so is performance sensitive. Could you make sure to add the results (before/after your changes) of the benchmarks provided here? https://github.com/DataDog/datadog-agent/blob/master/pkg/util/log/log_bench_test.go
* the `cisco_aci` integration file also has a `cert_key_password` field, is it already being cleaned up as expected? (might be worth adding a test case in any case)

KSerrania committed:
Fix log_test.go crash

KSerrania committed:
Add tests for cert_key_password

KSerrania commented:
@olivielpeau I added the benchmarks and the `cert_key_password` test case.

truthbk commented on pkg/util/log/strip.go:
```diff
@@ -68,6 +72,17 @@ func matchYAMLKey(key string) *regexp.Regexp {
 
+func matchCert() *regexp.Regexp {
+	/*
+	   Try to match as accurately as possible RFC 7468's ABNF
```
👌

truthbk commented on pkg/util/log/strip.go:
```diff
@@ -57,7 +56,12 @@ func init() {
 		Repl:  []byte(`$1 ********`),
 	}
-	replacers = []Replacer{apiKeyReplacer, appKeyReplacer, uriPasswordReplacer, passwordReplacer, tokenReplacer, snmpReplacer}
+	certReplacer = Replacer{
```
So my biggest concern here is we'll be running the regex, on every line logged (as called by the `scrub` function in the same package), this could drive CPU usage up quite a bit in a `debug` scenario. It would be nice to try to use a `Hint` to help keep this side-effect under control.

truthbk replied:
I jumped straight into the code not having have seen the previous review. I see we're trying to keep track of the performance impact 👌

KSerrania replied:
You're right about the `Hint`: I should add `BEGIN` (or `END`) as a hint so that it doesn't try to apply the real regexp if it isn't here. I'm going to try it and see if that improves the benchmarks (which should theoretically happen, since we don't put certificates in our benchs).
Note: should we add benchmarks where there can be certificates in the logs then?

KSerrania replied:
I also thought of there's another way to do this that might be less costly, but it's more convoluted and requires more work, so I don't know if that's worth the time.
Basically it would require creating a new type of replacer with a `StartRegex` and an `EndRegex`. We feed it the lines one by one: 
- if the line doesn't match the `StartRegex`, we write the line to the result
- if the line matches the `StartRegex`, we start keeping the lines (without writing the result) until we match the `EndRegex`. At this point, we replace everything that was between these two lines by whatever was set in the `Repl` or `ReplFunc` fields.
Moreover, we could add a `StartHint` and an `EndHint` for these two regexes (since we're matching by line).

truthbk reviewed (commented):
Really nice, clean stuff 🍰  

Added some feedback regarding potential side-effects, but let me know your thoughts @KSerrania 

Regarding the benchmarks, the most relevant benchmark I believe is `BenchmarkLogScrubbing` where we seem to have an increase of about 5.6% in ns/op. It's not too bad. I'm wondering if you see a way to harness `Hints` and see if we can make that a little better? These numbers are pretty good, but I'm curious if we could do a bit better. Being extra-cautious because this bit us once before.

(Still trying to understand how `BenchmarkLogScrubbingMulti` possibly went down 🤔)

KSerrania committed:
Add hint for certReplacer

KSerrania commented:
New benchmarks (with `Hint` added)

Before change:
```
goos: darwin
goarch: amd64
pkg: github.com/DataDog/datadog-agent/pkg/util/log
BenchmarkLogVanilla-8                1000000          2134 ns/op        1068 B/op         15 allocs/op
BenchmarkLogVanillaLevels-8         30000000            56.9 ns/op        64 B/op          2 allocs/op
BenchmarkLogScrubbing-8               100000         13231 ns/op        6597 B/op         38 allocs/op
BenchmarkLogScrubbingLevels-8       20000000            68.4 ns/op        16 B/op          1 allocs/op
BenchmarkLogScrubbingMulti-8          100000         15175 ns/op        7563 B/op         54 allocs/op
PASS
coverage: 26.3% of statements
```

After change:
```
goos: darwin
goarch: amd64
pkg: github.com/DataDog/datadog-agent/pkg/util/log
BenchmarkLogVanilla-8                1000000          2266 ns/op        1068 B/op         15 allocs/op
BenchmarkLogVanillaLevels-8         30000000            57.4 ns/op        64 B/op          2 allocs/op
BenchmarkLogScrubbing-8               100000         13462 ns/op        6680 B/op         39 allocs/op
BenchmarkLogScrubbingLevels-8       20000000            67.5 ns/op        16 B/op          1 allocs/op
BenchmarkLogScrubbingMulti-8          100000         15106 ns/op        7644 B/op         55 allocs/op
PASS
coverage: 27.8% of statements
```

KSerrania commented:
@truthbk A possible explanation for why some benchmarks went down is that I removed a thing that was strange: after applying the replacers to a line (which is a `[]byte`, we used to cast this line to a `string` and append it to a result string. And then when the entire file is taken care of, we re-cast this `string` into a `[]byte`.
Since there was no need to do these two casts (and that I needed a `[]byte` at the end of the replacers loop to apply my new replacer), I removed the intermediary `string` and kept everything in a `[]byte`.

truthbk reviewed (dismissed):
Great stuff, really happy with the implementation and performance :)

Sorry it took a while to get back to you! 🙇

truthbk reviewed (dismissed):
🚀

KSerrania committed:
Add release note
